src/Features/Core/Portable/PullMemberUp/AbstractCodeAndSolutionGenerator.cs

```diff
@@ -0,0 +1,336 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeGeneration;
+using Microsoft.CodeAnalysis.CodeRefactorings.PullMemberUp;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Formatting;
+using static Microsoft.CodeAnalysis.CodeActions.CodeAction;
+
+namespace Microsoft.CodeAnalysis.PullMemberUp
+{
+    internal abstract class AbstractCodeAndSolutionGenerator : IPullMemberUpActionAndSolutionGenerator
+    {
+        private Dictionary<DocumentId, DocumentEditor> EditorMap { get; set; }
+
+        private Dictionary<ISymbol, IEnumerable<SyntaxNode>> SyntaxMap { get; set; }
+
+        private ICodeGenerationService CodeGenerationService { get; set; }
+
+        public async Task<Solution> GetSolutionAsync(
+            AnalysisResult result,
+            Document contextDocument,
+            CancellationToken cancellationToken = default)
+        {
+            CodeGenerationService = contextDocument.Project.LanguageServices.GetRequiredService<ICodeGenerationService>();
+            var solution = contextDocument.Project.Solution;
+            var targetNodeSyntax = await CodeGenerationService.FindMostRelevantNameSpaceOrTypeDeclarationAsync(solution, result.Target).ConfigureAwait(false);
+
+            if (targetNodeSyntax != null)
+            {
+                var solutionEditor = new SolutionEditor(solution);
+                await InitializeEditorsAndSyntaxes(result, targetNodeSyntax, solutionEditor, solution, cancellationToken).ConfigureAwait(false);
+                if (result.Target.TypeKind == TypeKind.Interface)
+                {
+                    return GenerateTargetingInterfaceSolution(result, targetNodeSyntax, solutionEditor, solution);
+                }
+                else if (result.Target.TypeKind == TypeKind.Class)
+                {
+                    return GenerateTargetingClassSolution(result, targetNodeSyntax, solutionEditor, solution);
+                }
+                else
+                {
+                    throw new ArgumentException($"{nameof(result.Target)} should be interface or class");
+                }
+            }
+            else
+            {
+                return default;
+            }
+        }
+
+        public CodeAction GetCodeActionAsync(
+            AnalysisResult result,
+            Document contextDocument,
+            string title)
+        {
+            if (!result.IsValid)
+            {
+                return default;
+            }
+
+            CodeGenerationService = contextDocument.Project.LanguageServices.GetRequiredService<ICodeGenerationService>();
```

Comment:
you're mutating state of values that can be shared, that's not ok.   this also seems unnecessary since you're passing context along.